File: frontend/src/app/core/models/income.model.ts
export type IncomeType =
  | 'salary'
  | 'freelance'
  | 'bonus'
  | 'commission'
  | 'gift'
  | 'dividend'
  | 'business'
  | 'one_time'
  | 'custom';

export type RecurrenceType =
  | 'daily'
  | 'weekly'
  | 'monthly'
  | 'quarterly'
  | 'yearly'
  | 'one_time'
  | 'custom';

export type IncomeStatus = 'received' | 'pending' | 'scheduled' | 'cancelled';

export type PaymentMode =
  | 'bank_transfer'
  | 'cash'
  | 'card'
  | 'upi'
  | 'paypal'
  | 'crypto'
  | 'cheque'
  | 'other';

export interface CustomField {
  key: string;
  value: unknown;
  label?: string;
}

export interface AttachmentMeta {
  file_name: string;
  url?: string;
  content_type?: string;
  size?: number;
}

export interface Income {
  id: string;
  user_id: string;
  title: string;
  amount: number;
  currency: string;
  income_type: IncomeType;
  recurrence: RecurrenceType;
  status: IncomeStatus;
  payment_date: string;
  start_date?: string | null;
  end_date?: string | null;
  category_id?: string | null;
  source_id?: string | null;
  source_name?: string | null;
  tag_ids: string[];
  payment_mode?: PaymentMode | null;
  platform?: string | null;
  client?: string | null;
  employer?: string | null;
  project_name?: string | null;
  reference_number?: string | null;
  notes?: string | null;
  tax_notes?: string | null;
  is_taxable: boolean;
  attachments: AttachmentMeta[];
  custom_fields: CustomField[];
  metadata: Record<string, unknown>;
  created_at?: string | null;
  updated_at?: string | null;
}

export type IncomePayload = Partial<Omit<Income, 'id' | 'user_id' | 'created_at' | 'updated_at'>> & {
  title: string;
  amount: number;
  payment_date: string;
};

export interface IncomeFilters {
  search?: string | null;
  income_type?: IncomeType | null;
  recurrence?: RecurrenceType | null;
  status?: IncomeStatus | null;
  payment_mode?: PaymentMode | null;
  category_id?: string | null;
  source_id?: string | null;
  tag_id?: string | null;
  currency?: string | null;
  min_amount?: number | null;
  max_amount?: number | null;
  start_date?: string | null;
  end_date?: string | null;
  is_recurring?: boolean | null;
}

File: backend/app/modules/incomes/models/income_model.py
"""Internal domain model describing an income document."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.core.constants import (
    IncomeStatus,
    IncomeType,
    PaymentMode,
    RecurrenceType,
)


class IncomeModel(BaseModel):
    """Shape of an income document stored in MongoDB.

    Designed for flexibility: a free-form ``metadata`` map and
    ``custom_fields`` list let users extend the schema without migrations.
    """

    user_id: str
    title: str
    amount: float
    currency: str = "USD"
    income_type: IncomeType = IncomeType.SALARY
    recurrence: RecurrenceType = RecurrenceType.ONE_TIME
    status: IncomeStatus = IncomeStatus.RECEIVED

    payment_date: datetime
    payment_time: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None

    # Recurring-income automation.
    day_of_month: Optional[int] = None
    auto_add: bool = False
    is_auto_generated: bool = False
    recurring_parent_id: Optional[str] = None
    next_run_at: Optional[datetime] = None

    category_id: Optional[str] = None
    source_id: Optional[str] = None
    source_name: Optional[str] = None
    tag_ids: List[str] = Field(default_factory=list)

    payment_mode: Optional[PaymentMode] = None
    platform: Optional[str] = None
    client: Optional[str] = None
    employer: Optional[str] = None
    project_name: Optional[str] = None
    reference_number: Optional[str] = None

    notes: Optional[str] = None
    tax_notes: Optional[str] = None
    is_taxable: bool = True

    attachments: List[Dict[str, Any]] = Field(default_factory=list)
    custom_fields: List[Dict[str, Any]] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(default_factory=dict)

    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
